src/main/java/com/google/sps/servlets/MealServlet.java

```diff
@@ -105,29 +116,56 @@ private void getMealById(HttpServletRequest request, HttpServletResponse respons
             response.sendError(HttpServletResponse.SC_NOT_FOUND);
             return;
         }
-
-        Meal meal = dishes.get(id);
-
-        if (meal != null) {
-            String gson = new Gson().toJson(meal);
-            response.setContentType("application/json");
-            response.getWriter().println(gson);
-            return;
+        Filter propertyFilter = new FilterPredicate("id", FilterOperator.EQUAL, id);
+        Query query = new Query("Meal").setFilter(propertyFilter);
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+        ArrayList<Meal> meals = getDataFromDatastore(results); 
+        
+        if (meals.size() > 1) {
+            response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+        }
+        if (meals.size() == 1) {
+            Meal meal = meals.get(0);
+            if (meal != null) {
+                String gson = new Gson().toJson(meal);
+                response.setContentType("application/json");
+                response.getWriter().println(gson);
+                return;
+            }
         }
-
         response.sendError(HttpServletResponse.SC_NOT_FOUND);
         return;
     }
 
-     private void returnIdOfSimilar(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    private ArrayList<Meal> getDataFromDatastore(PreparedQuery query) {
+        ArrayList<Meal> result = new ArrayList<>();
+        for (Entity entity : query.asIterable()) {
+            Long id = (Long)entity.getProperty("id");
+            String title = (String) entity.getProperty("title");
+            String description = (String) entity.getProperty("description");
+            ArrayList<String> ingredients = (ArrayList<String>) entity.getProperty("ingredients");
+            String type = (String) entity.getProperty("type");
```

Comment:
You need to put all this in a `try {} catch` block, in case a `ClassCastException` happens (because a field may not convert to the type you expect).
You also need to test what happens if one of these fields is not present, or is empty (such as type="", ingredients being an empty list, etc).